### PR TITLE
add synopsys code scanning workflow

### PR DESCRIPTION
**Pull request by rahulguna10**
This PR adds a new Code Scanning template workflow for Synopsys Intelligent Security Scan Action.

The Synopsys Intelligent Security Scan Action helps selectively perform SAST and SCA scans, triggered during a variety of GitHub Platform events, such as push or pull request. The Synopsys Intelligent Security Scan Action allows your projects to run the only required type of security scans, optimizing the time taken by security testing and provide quicker feedback on scan results.

Prerequisites
- To use this Action you must be a licensed Polaris customer.
- Intelligent scan server must be deployed and accessible via GitHub Actions.

---

### **Please note that at this time we are only accepting new starter workflows for Code Scanning. Updates to existing starter workflows are fine.**

---

## Tasks

**For _all_ workflows, the workflow:**

- [x] Should be contained in a `.yml` file with the language or platform as its filename, in lower, [_kebab-cased_](https://en.wikipedia.org/wiki/Kebab_case) format (for example, [`docker-image.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-image.yml)).  Special characters should be removed or replaced with words as appropriate (for example, "dotnet" instead of ".NET").
- [x] Should use sentence case for the names of workflows and steps (for example, "Run tests").
- [x] Should be named _only_ by the name of the language or platform (for example, "Go", not "Go CI" or "Go Build").
- [x] Should include comments in the workflow for any parts that are not obvious or could use clarification.

**For _Code Scanning_ workflows, the workflow:**

- [x] Should be preserved under [the `code-scanning` directory](https://github.com/actions/starter-workflows/tree/main/ci).
- [x] Should include a matching `code-scanning/properties/*.properties.json` file (for example, [`code-scanning/properties/codeql.properties.json`](https://github.com/actions/starter-workflows/blob/main/code-scanning/properties/codeql.properties.json)), with properties set as follows:
  - [x] `name`: Name of the Code Scanning integration.
  - [x] `organization`: Name of the organization producing the Code Scanning integration.
  - [x] `description`: Short description of the Code Scanning integration.
  - [x] `categories`: Array of languages supported by the Code Scanning integration.
  - [x] `iconName`: Name of the SVG logo representing the Code Scanning integration. This SVG logo must be present in [the `icons` directory](https://github.com/actions/starter-workflows/tree/main/icons).
- [x] Should run on `push` to `branches: [ $default-branch, $protected-branches ]` and `pull_request` to `branches: [ $default-branch ]`. We also recommend a `schedule` trigger of `cron: $cron-weekly` (for example, [`codeql.yml`](https://github.com/actions/starter-workflows/blob/c59b62dee0eae1f9f368b7011cf05c2fc42cf084/code-scanning/codeql.yml#L14-L21)).

**Some general notes:**

- ~This workflow must _only_ use actions that are produced by GitHub, [in the `actions` organization](https://github.com/actions),~ **or**
- [x] This workflow must _only_ use actions that are produced by the language or ecosystem that the workflow supports.  These actions must be [published to the GitHub Marketplace](https://github.com/marketplace?type=actions).  We require that these actions be referenced using the full 40 character hash of the action's commit instead of a tag.  Additionally, workflows must include the following comment at the top of the workflow file:
    ```
    # This workflow uses actions that are not certified by GitHub.
    # They are provided by a third-party and are governed by
    # separate terms of service, privacy policy, and support
    # documentation.
    ```

### CODE REVIEW

File: code-scanning/properties/synopsysio.properties.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+    "name": "Synopsys Intelligent Security Scan Action",
+    "creator": "Synopsys",
+    "description": "The Synopsys Intelligent Security Scan Action helps selectively perform SAST and SCA scans, triggered during a variety of GitHub Platform events for C, C++, C#, Java, JavaScript, TypeScript, Python, and Go developers.",
+    "iconName": "synopsys",
```
[the remark this answers is not in the document]
Comment (rahulguna10) [reply code-scanning/properties/synopsysio.properties.json:5]: When I change the iconName to synopsys-io, do I also need to rename the synopsys.svg to synopsys-io.svg ?

[the remark this answers is not in the document]
Comment (nickfyson) [reply code-scanning/properties/synopsysio.properties.json:5]: Yes, that's right.

Basically, for ease of understanding as we add more workflows all the files you're adding here should have the same name, eg....
- `synopsys-io.svg`
- `synopsys-io.yml`
- `synopsys-io.properties.json`

[the remark this answers is not in the document]
Comment (rahulguna10) [reply code-scanning/properties/synopsysio.properties.json:5]: Got it.

[the remark this answers is not in the document]
Comment (rahulguna10) [reply code-scanning/properties/synopsysio.properties.json]: I have pushed the changes @nickfyson

File: code-scanning/synopsys-io-actions.yml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    name: Analyze
+    runs-on: ubuntu-latest
+    
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v2
+      
+    - name: Synopsys Intelligent Security Scan
+      id: prescription
+      uses: synopsys-sig/intelligent-security-scan@v1
```
Comment (nickfyson) [inline code-scanning/synopsys-io-actions.yml:28]:
```suggestion
      uses: synopsys-sig/intelligent-security-scan@6940a88f5a0062304657dd62e7d3cee3b5bc9e57
```

File: code-scanning/synopsys-io-actions.yml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    # in order for this logic to work
+    - name: Software Composition Analysis with Black Duck
+      if: ${{steps.prescription.outputs.scaScan == 'true' }}
+      uses: blackducksoftware/github-action@v2
+      with:
+         args: '--blackduck.url="${{ secrets.BLACKDUCK_URL}}" --blackduck.api.token="${{ secrets.BLACKDUCK_TOKEN}}" --detect.tools="SIGNATURE_SCAN,DETECTOR"'
+    
+    - name: Synopsys Intelligent Security Scan
+      if: ${{ steps.prescription.outputs.sastScan == 'true' || steps.prescription.outputs.scaScan == 'true' }}
+      uses: synopsys-sig/intelligent-security-scan@v1
```
Comment (nickfyson) [inline code-scanning/synopsys-io-actions.yml:58]:
```suggestion
      uses: synopsys-sig/intelligent-security-scan@6940a88f5a0062304657dd62e7d3cee3b5bc9e57
```

File: code-scanning/synopsys-io-actions.yml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+          export POLARIS_ACCESS_TOKEN=${{ secrets.POLARIS_ACCESS_TOKEN}}
+          wget -q ${{ secrets.POLARIS_SERVER_URL}}/api/tools/polaris_cli-linux64.zip
+          unzip -j polaris_cli-linux64.zip -d /tmp
+          /tmp/polaris analyze -w
+    
+    # Please note that the ID in previous step was set to prescription
+    # in order for this logic to work
+    - name: Software Composition Analysis with Black Duck
+      if: ${{steps.prescription.outputs.scaScan == 'true' }}
+      uses: blackducksoftware/github-action@v2
```
Comment (nickfyson) [inline code-scanning/synopsys-io-actions.yml:52]:
```suggestion
      uses: blackducksoftware/github-action@9ea442b34409737f64743781e9adc71fd8e17d38
```

**Review by nickfyson — CHANGES_REQUESTED** [review]
Thanks for these updates!

The last thing we need is to use the full SHA rather than a tag to specify the workflow version, as that's a security requirement for actions that aren't under GitHub's direct control. I took the liberty of looking up the corresponding SHAs and adding them as suggested changes, but you might want to check for typos! 😆 

Then we'll be good to go! 🚀

File: code-scanning/synopsys-io-actions.yml (hunk 4 above)
Comment (nickfyson) [reply code-scanning/synopsys-io-actions.yml:52]: I think this is the appropriate SHA for this action? 🙂

**Review by nickfyson — APPROVED** [review]
LGTM 🙂

Thanks for the new workflow, and for jumping through all the (alas necessary) hoops! 😉

**Pull request merged**